Adapt latest java-libkiwix wrapper

MohitMaliFtechiz opened the pull request:
Fixes #3216

kelson42 commented:
@MohitMaliFtechiz Does this PR fixes other tickezs in 0.8 milestone? If "yes", which one exactly?

MohitMaliFtechiz commented:
@kelson42, Yes it seems to fix some other tickets of milestone `3.8.0`.
1) https://github.com/kiwix/kiwix-android/issues/3094
2) i) https://github.com/kiwix/kiwix-android/issues/3410
    ii) https://github.com/kiwix/kiwix-android/issues/2768 Once searching is start working we will test these issues.
3) https://github.com/kiwix/kiwix-android/issues/3442 This issue is fixed in the latest wrapper.

Some tickets would be fixed by this PR, but not tested yet.
1) https://github.com/kiwix/kiwix-android/issues/2767

kelson42 commented:
So why they are not in description?

MohitMaliFtechiz commented:
@gouri-panda, This PR is ready for review now. But here we have fixed `SuggestionSearch` in `java-libkiwix` https://github.com/kiwix/java-libkiwix/pull/61. So once we will publish the new version of `java-libkiwix` on Maven then we upgrade the version of libkiwix and we can merge this PR meanwhile you can start reviewing this PR since it is a big PR.

gouri-panda commented:
@MohitMaliDeveloper Ok, but it'll take some time to review the code since I'm not quite familiar with this end of the code and the PR is large and it also touches breakable changes. I want to make things right :)

kelson42 commented:
@MohitMaliFtechiz we need anyway to republish the binding with the right name first. please prepare everything for the release in a pr (see release ticket in java-libkiwix)

kelson42 commented:
@gouri-panda Any news here? We really need your expert approval here to move forward!

kelson42 commented:
org.kiwix/libkiwix has been released on Maven. Please update PR to use it and review with PRIO0.

MohitMaliFtechiz commented:
> org.kiwix/libkiwix has been released on Maven. Please update PR to use it and review with PRIO0.

@kelson42 I have made the changes, Now we are using the new libkiwix from Maven.

kelson42 commented:
@MohitMaliFtechiz API24 still failing, plrase fix in top prio.

MohitMaliFtechiz commented:
@kelson42, @gouri-panda I have improved the test cases for better performance and fixed the memory leak in the application which caused the test failure also updated the PR description what was the problem and what we have done to fix the issue.

kelson42 commented:
@MohitMaliFtechiz Why the codecov task fails after 0 seconds?

MohitMaliFtechiz commented:
> @MohitMaliFtechiz Why the codecov task fails after 0 seconds?

@kelson42 `codecov/project` Ci is running but `codecov/patch` CI is failing. When there is no change in the patch report then this CI fails like [Here](https://github.com/kiwix/kiwix-android/pull/3473/checks?check_run_id=16340766113) because we have set the patch target to 70%. As you can see [here](https://github.com/kiwix/kiwix-android/pull/3455/checks?check_run_id=15967441001)  `codecov/project` succeeds in 0 seconds because there is no change in the report but we have not set any target in `project` so it passed otherwise it also fails.

MohitMaliFtechiz commented:
> @MohitMaliFtechiz please rebase and fix conflict

@kelson42 Done.

kelson42 commented:
@gouri-panda Any news here?

kelson42 commented:
@gouri-panda Unfortunately, i will have to merge without your expert review if nothing as moved on until next week.

gouri-panda commented on core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchFragment.kt:
```diff
@@ -104,13 +113,71 @@ class SearchFragment : BaseFragment() {
+          val totalItemCount = layoutManager.itemCount
+          val lastVisibleItem = layoutManager.findLastVisibleItemPosition()
+          // Check if the user is about to reach the last item
+          if (!isDataLoading && totalItemCount <= lastVisibleItem + VISIBLE_ITEMS_THRESHOLD) {
+            // Load more data when the last item is almost visible
+            loadMoreSearchResult()
```
Just an idea, currently we are loading the search results if user is in the last items. What if start the loading before the 3 last items. 

```
 if (!isDataLoading && totalItemCount  <= lastVisibleItem + VISIBLE_ITEMS_THRESHOLD - 3) {
```
What do you think @MohitMaliDeveloper ?

MohitMaliFtechiz replied:
I had this in mind but I forgot to implement this my bad.

MohitMaliFtechiz replied:
Done

gouri-panda commented on core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchFragment.kt:
```diff
@@ -104,13 +113,71 @@ class SearchFragment : BaseFragment() {
+   * If there are no new non-duplicate results, the method sets `isDataLoading` to true
+   * to avoid unnecessary data loading, as there are no more items available.
+   */
+  private fun loadMoreSearchResult() {
+    if (isDataLoading) return
+    val safeStartIndex = searchAdapter?.itemCount ?: 0
+    isDataLoading = true
```
We should extract this data logic to repostiory classes to make separate concerns.

MohitMaliFtechiz replied:
Sorry, I didn't get you, are you talking about `val safeStartIndex = searchAdapter?.itemCount ?: 0` this?

gouri-panda replied:
No, the logic of the function shouldn't be in the `SearchFragment`. We should refactor to the data classes maybe viewmodel or repository depending on the situation for separate concerns.

MohitMaliFtechiz replied:
@gouri-panda Done.

gouri-panda commented on core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchFragment.kt:
```diff
@@ -104,13 +113,71 @@ class SearchFragment : BaseFragment() {
+    isDataLoading = true
+    fragmentSearchBinding?.loadingMoreDataIndicator?.isShowing(true)
+
+    CoroutineScope(Dispatchers.IO).launch {
+      val fetchMoreSearchResults = searchState?.getVisibleResults(safeStartIndex)
+      CoroutineScope(Dispatchers.Main).launch {
+        fragmentSearchBinding?.loadingMoreDataIndicator?.isShowing(false)
+        isDataLoading = when {
```
We should use one scope to manage(add/cancel etc)) coroutines i.e. `viewModelScope`.

MohitMaliFtechiz replied:
Done

gouri-panda commented on core/src/main/java/org/kiwix/kiwixmobile/core/search/SearchFragment.kt:
```diff
@@ -194,11 +264,24 @@ class SearchFragment : BaseFragment() {
-    fragmentSearchBinding?.searchLoadingIndicator?.isShowing(state.isLoading)
-    fragmentSearchBinding?.searchNoResults?.isVisible = state.visibleResults.isEmpty()
-    searchAdapter?.items = state.visibleResults
+    fragmentSearchBinding?.searchLoadingIndicator?.isShowing(true)
+    renderingJob = CoroutineScope(Dispatchers.Main).launch {
+      val searchResult = withContext(Dispatchers.IO) {
```
Use `ViewModelScope` and use `withContext` to switch threads.

MohitMaliFtechiz commented on app/src/androidTest/java/org/kiwix/kiwixmobile/search/SearchRobot.kt:
```diff
@@ -32,7 +37,12 @@ class SearchRobot : BaseRobot() {
   fun clickOnSearchItemInSearchList() {
     BaristaSleepInteractions.sleep(TestUtils.TEST_PAUSE_MS_FOR_SEARCH_TEST.toLong())
     isVisible(ViewId(R.id.search_list))
-    clickOn(ViewId(R.id.list_item_search_text))
+    onView(withId(R.id.search_list)).perform(
+      actionOnItemAtPosition<RecyclerView.ViewHolder>(
+        0,
+        click()
```
@gouri-panda, Yes there is a difference, we are clicking on the first item of the `recycler view` since we are using the `SuggestionSearch` so it gives us more results, and here we want to click on the first item to test `search` functionality. This approach is good for future test cases if we change the zim file and if that gives results more than one then it will only click on the first item of that search list to properly test the functionality. In the previous test case, we assumed there was only one result coming from the search and we were directly clicking on the `recycler view item id`.

MohitMaliFtechiz commented on .github/workflows/ci.yml:
```diff
@@ -79,16 +79,16 @@ jobs:
 
       - name: create unit coverage
-        if: ${{ matrix.api-level==21 }}
+        if: ${{ matrix.api-level==24 }}
```
@gouri-panda Last time we placed this condition as @kelson42 requested, `Unit coverage` works similarly for every API level and we are testing all the UI functionality in `instrumentation test cases` for every API level, so is it beneficial to run `unit coverage` for every API level?

kelson42 replied:
its would be too much for code coverage only.

gouri-panda replied:
@MohitMaliDeveloper but we are here running this `run: ./gradlew testDebugUnitTest testCustomexampleDebugUnitTest`

MohitMaliFtechiz replied:
@gouri-panda We are running unit tests for both the custom and app modules i think this naming is good since it creates the coverage for the whole project, if you have a better suggestion for this please share your thoughts.

MohitMaliFtechiz commented on app/src/main/java/org/kiwix/kiwixmobile/webserver/KiwixServer.kt:
```diff
@@ -19,26 +19,37 @@
 private const val TAG = "KiwixServer"
 
-class KiwixServer @Inject constructor(private val jniKiwixServer: JNIKiwixServer) {
+// jniKiwixServer is a server running on a existing library.
+// We must keep the library alive (keep a reference to it) to not delete the library the server
+// is working on.
```
Here `Library` holds the books(Zim files) that we host on the `Server` (Kiwix Server) so for hosting books on the server we need to keep the library object otherwise it will crash the server because of library object is not alive.

Full conversation https://github.com/kiwix/java-libkiwix/issues/51

MohitMaliFtechiz replied:
Done

MohitMaliFtechiz commented on app/src/main/java/org/kiwix/kiwixmobile/webserver/KiwixServer.kt:
```diff
@@ -19,26 +19,37 @@
       selectedBooksPath.forEach { path ->
         try {
-          kiwixLibrary.addBook(path)
+          val book = Book().apply {
```
```
val book = Book().apply {
            update(Archive(path))
          }
```

@gouri-panda Here we are using the new wrapper and it has many changes, so we need the `Book` object for the `Library` here we are creating an empty book object first and using the `update` function of the `Book` class to add the `Archive` in it.

MohitMaliFtechiz commented on app/src/androidTest/java/org/kiwix/kiwixmobile/mimetype/MimeTypeTest.kt:
```diff
@@ -70,12 +76,12 @@ class MimeTypeTest : BaseActivityTest() {
     }
     val zimFileReader = ZimFileReader(
       zimFile,
-      JNIKiwixReader(zimFile.canonicalPath),
+      Archive(zimFile.canonicalPath),
```
@gouri-panda Here `Archive` works like `Reader` we were previously using, it has all the data about the `Zim file` e.g. illustration, metadata, main page entry etc. you can see all methods of `Archive` class here  https://github.com/kiwix/java-libkiwix/blob/main/lib/src/main/java/org/kiwix/libzim/Archive.java

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/reader/ZimFileReader.kt:
```diff
@@ -127,8 +158,8 @@ class ZimFileReader constructor(
 
-  fun readContentAndMimeType(uri: String): String = getContentAndMimeType(uri)
-    .second.truncateMimeType.also {
+  fun getMimeTypeFromUrl(uri: String): String? = getItem(uri)?.mimetype
+    ?.truncateMimeType.also {
```
Good point, I agree with you :)

MohitMaliFtechiz replied:
I have added test cases for it.

MohitMaliFtechiz commented on core/src/main/java/org/kiwix/kiwixmobile/core/reader/ZimFileReader.kt:
```diff
@@ -72,18 +83,39 @@ class ZimFileReader constructor(
+      try {
+        jniKiwixReader.mainEntry.getItem(true).path
+      } catch (ignore: Exception) {
+        null
```
Done

gouri-panda reviewed (commented):
All changes look good! We have a little bit more discussion about your previous comments.  I tested the app yesterday. So far, looks good. After a final review,  we are good to merge!
